[Docs] Data model (#1374)
wasp-lang/wasp: web/docs/database/operations/actions.md

---
title: Actions
---

import { Required } from '@site/src/components/Required';

We'll explain what Actions are and how to use them. If you're looking for a detailed API specification, skip ahead to the [Options Reference](#options-reference).

> Actions are very similar to [Queries](/docs/database/operations/queries.md). The key difference is that Actions are meant to modify and add data, while Queries are only supposed to read it. Adding a comment to a blog post, liking a video, updating a product's price... These are some common examples of Actions.
> 
> Actions and Queries work hand-in-hand on keeping data caches fresh.

Reviewer: ```suggestion
Actions are quite similar to [Queries](/docs/database/operations/queries.md), but with a key distinction: Actions are designed to modify and add data, while Queries are solely for reading data. Examples of Actions include adding a comment to a blog post, liking a video, or updating a product's price.

Actions and Queries work together to keep data caches up-to-date.
```

:::tip
Actions are almost identical to Queries in terms of their API.
Therefore, if you're already familiar with Queries, you might find reading the entire guide repetitive.

We instead recommend skipping ahead and only reading [the differences between Queries and Actions](#differences-between-queries-and-actions), and consulting the [Options Reference](#options-reference) as needed.
:::

> ## Working with Actions
> 
> Actions are declared in Wasp and implemented in NodeJS. Wasp executes Actions within the server's context but also generates the code that lets you call them from anywhere in your code (client or server) via the same interface.
> In other words, you don't have to worry about building an HTTP API for the Action, handling the request on the server, or even handling and caching the responses on the client.
> Instead, simply focus on the business logic inside your Action and let Wasp take care of the rest!
> 
> To create an Action, you must:
> 
> 1. Declare the Action in Wasp using the `action` declaration.
> 2. Define the Action's NodeJS implementation.
> 
> After completing these two steps, you'll be able to use the Action from any point in your code.

Reviewer: ```suggestion
## Working with Actions

Actions are declared in Wasp and implemented in NodeJS. Wasp runs Actions within the server's context, but it also generates code that allows you to call them from anywhere in your code (either client or server) using the same interface. This means you don't have to worry about building an HTTP API for the Action, managing server-side request handling, or even dealing with client-side response handling and caching.

Instead, just focus on developing the business logic inside your Action, and let Wasp handle the rest!

To create an Action, you need to:

1. Declare the Action in Wasp using the `action` declaration.
2. Implement the Action's NodeJS functionality.

Once these two steps are completed, you can use the Action from anywhere in your code.
```

> ### Declaring Actions
> 
> We'll start by telling Wasp we wish to create an Action using an `action` declaration.
> Let's declare two Actions, one for creating a task and another one for marking tasks as done:

Reviewer: ```suggestion
### Declaring Actions

To create an Action in Wasp, we begin with an `action` declaration. Let's declare two Actions - one for creating a task, and another for marking tasks as done:
```

action createTask {
  fn: import { createTask } from "@server/actions.js"
}

action markTaskAsDone {
  fn: import { markTaskAsDone } from "@server/actions.js"
}

```

</TabItem>
<TabItem value="ts" label="TypeScript">

```wasp title="main.wasp"
// ...

action createTask {
  fn: import { createTask } from "@server/actions.js"
}

action markTaskAsDone {
  fn: import { markTaskAsDone } from "@server/actions.js"
}
```

:::warning
Even though you are using TypeScript and plan to implement this Action in `src/server/actions.ts`, you still need to import it using a `.js` extension. Wasp internally uses `esnext` module resolution, which requires importing all files with a `.js` extension. This is only needed when importing `@server` files.

Read more about ES modules in TypeScript [here](https://www.typescriptlang.org/docs/handbook/esm-node.html). If you're interested in the discussion and the reasoning behind this, read about it [in this GitHub issue](https://github.com/microsoft/TypeScript/issues/33588).
:::

</TabItem>
</Tabs>

<small>

If you want to know about all supported options for the `action` declaration, take a look at the [Options reference](#options-reference).

</small>

> Wasp Actions and their implementations don't need to (but can) have the same name. We'll keep the names identical to avoid confusion.

Reviewer: ```suggestion
The names of Wasp Actions and their implementations don't necessarily have to match. However, to avoid confusion, we'll keep them the same.
```

:::info
You might have noticed that we told Wasp to import Action implementations that don't yet exist. Don't worry about that for now. We'll write the implementations imported from `actions.{js,ts}` in the next section.

It's a good idea to start with the high-level concept (i.e., the Action declaration in the Wasp file) and only then deal with the implementation details (i.e., the Action's implementation in JavaScript).
:::

> After declaring a Wasp Action, two crucial things happen:
> 
> - Wasp **generates a server-side NodeJS function** that shares its name with the Action. This function's interface is identical to the client-side function from the previous point.
> - Wasp **generates a client-side JavaScript function** that shares its name with the Action (e.g., `createTask`).
>   This function takes a single optional argument - an object containing any serializable data you wish to use inside the Action.
>   Wasp will send this object over the network and pass it into the Action's implementation as its first positional argument (more on this when we look at the implementations).
>   Such an abstraction works thanks to an HTTP API route handler Wasp generates on the server, which calls the Action's NodeJS implementation under the hood.
> 
> Generating these two functions ensures a uniform calling interface across the entire app (both client and server).

Reviewer: ```suggestion
After declaring a Wasp Action, two key things occur:

- Wasp **creates a server-side NodeJS function** that has the same name as the Action.
  This function's interface mirrors that of the client-side function from the previous point.
- Wasp **produces a client-side JavaScript function** that also shares the Action's name (e.g., `createTask`).

  This function accepts a single optional argument - an object containing any serializable data you want to use inside the Action. Wasp will transmit this object over the network and pass it into the Action's implementation as its first positional argument (we'll discuss this more when we look at the implementations). This abstraction is facilitated by an HTTP API route handler that Wasp generates on the server, which calls the Action's NodeJS implementation behind the scenes.

These two generated functions ensure a consistent calling interface across your entire app, whether it's on the client or server side.
```

### Implementing Actions in Node

> Now that we've declared the Action, all that's left is implementing it.
> We've told Wasp to look for Actions' implementations in the file `src/server/actions.{js,ts}`, so that's where we should export them from.
> 
> Here's how you might implement the previously declared Actions `createTask` and `markTaskAsDone`:

Reviewer: ```suggestion
Now that we've declared the Action, what remains is to implement it. We've instructed Wasp to look for the Actions' implementations in the file `src/server/actions.{js,ts}`, so that's where we should export them from.

Here's how you might implement the previously declared Actions `createTask` and `markTaskAsDone`:
```

<TabItem value="js" label="JavaScript">

```js title="src/server/actions.js"
// our "database"
let nextId = 4;
const tasks = [
  { id: 1, description: "Buy some eggs", isDone: true },
  { id: 2, description: "Make an omelette", isDone: false },
  { id: 3, description: "Eat breakfast", isDone: false },
];

// You don't need to use the arguments if you don't need them
export const createTask = (args) => {
  const newTask = {
    id: nextId,
    isDone: false,
    description: args.description,
  };
  nextId += 1;
  tasks.push(newTask);
  return newTask;
};

// The 'args' object is something sent by the caller (most often from the client)
export const markTaskAsDone = (args) => {
  const task = tasks.find((task) => task.id === args.id);
  if (!task) {
    // We'll show how to properly handle such errors later
    return;
  }
  task.isDone = true;
};
```

:::tip
Wasp uses [superjson](https://github.com/blitz-js/superjson) under the hood. In other words, you don't need to limit yourself to only sending and receiving JSON payloads.

Send and receive any superjson-compatible payload (e.g., Dates, Sets, Lists, circular references, etc.) and let Wasp take care of the (de)serialization.

> :::

Reviewer: This could be a "partial template" aka `_superjsonNote.md`

      </p>
      {isDone || (
        <button onClick={() => markTaskAsDone({ id })}>Mark as done.</button>
      )}
    </div>
  );
};
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```tsx {4,25} title=src/client/pages/Task.tsx
import React from "react";
import { useQuery } from "@wasp/queries";
import getTask from "@wasp/queries/getTask";
import markTaskAsDone from "@wasp/actions/markTaskAsDone";

export const TaskPage = ({ id }: { id: number }) => {
  const { data: task } = useQuery(getTask, { id });

  if (!task) {
    return <h1>"Loading"</h1>;
  }

  const { description, isDone } = task;
  return (
    <div>
      <p>
        <strong>Description: </strong>
        {description}
      </p>
      <p>
        <strong>Is done: </strong>
        {isDone ? "Yes" : "No"}
      </p>
      {isDone || (
        <button onClick={() => markTaskAsDone({ id })}>Mark as done.</button>
      )}
    </div>
  );
};
```

</TabItem>
</Tabs>

Since Actions don't require reactivity, they are safe to use inside components without a hook. Still, Wasp provides comes with the `useAction` hook you can use to enhance actions. Read all about it in the [Options Reference](#options-reference).

> ### Error Handling

Reviewer: I would maybe put this section, using entities and Prisma Error Helpers into the `Overview` section (since they are the same for actions and queries). 

I did something similar with the Social Auth docs 🤷

Author: Discussed it, postponed it.

Here's how you can import and use them:

<Tabs groupId="js-ts">
<TabItem value="js" label="JavaScript">

```js
import { isPrismaError, prismaErrorToHttpError } from "@wasp/utils.js";

// ...

try {
  await context.entities.Task.create({...})
} catch (e) {
  if (isPrismaError(e)) {
    throw prismaErrorToHttpError(e)
  } else {
    throw e
  }
}
```

</TabItem>
<TabItem value="ts" label="TypeScript">

```js
import { isPrismaError, prismaErrorToHttpError } from "@wasp/utils.js";

// ...

try {
  await context.entities.Task.create({...})
} catch (e) {
  if (isPrismaError(e)) {
    throw prismaErrorToHttpError(e)
  } else {
    throw e
  }
}
```

</TabItem>
</Tabs>

## Cache Invalidation

One of the trickiest parts of managing a web app's state is making sure the data returned by the Queries is up to date.
Since Wasp uses _react-query_ for Query management, we must make sure to invalidate Queries (more specifically, their cached results managed by _react-query_) whenever they become stale.

It's possible to invalidate the caches manually through several mechanisms _react-query_ provides (e.g., refetch, direct invalidation).

> However, since manual cache invalidation quickly becomes complex and error-prone, Wasp offers a quicker and a more effective solution to get you started: **automatic Entity-based Query cache invalidation**.
> Because Actions can (and most often do) modify the state while Queries read it, Wasp invalidates a Query's cache whenever an Action that uses the same Entity is executed.

Reviewer: ```suggestion
To simplify this process, Wasp offers a faster and more effective solution: **automatic Entity-based Query cache invalidation**. Since Actions often modify the state while Queries read it, Wasp automatically invalidates a Query's cache whenever an Action that uses the same Entity is executed.
```

For example, let's assume that Action `createTask` and Query `getTasks` both use Entity `Task`. If `createTask` is executed, `getTasks`'s cached result may no longer be up-to-date.
Wasp will therefore invalidate it, making `getTasks` refetch data from the server, bringing it up to date again.

In practice, this means that Wasp keeps the Queries "fresh" without requiring you to think about cache invalidation.

On the other hand, this kind of automatic cache invalidation can become wasteful (some updates might not be necessary) and will only work for Entities. If that's an issue, you can use the mechanisms provided by _react-query_ for now, and expect more direct support in Wasp for handling those use cases in a nice, elegant way.

If you wish to optimistically set cache values after performing an Action, you can do so using [optimistic updates](https://stackoverflow.com/a/33009713). Configure them using Wasp's [useAction hook](#the-useaction-hook). This is currently the only manual cache invalidation mechanism Wasps supports natively. For everything else, you can always rely on _react-query_.

## Differences Between Queries and Actions

> Actions and Queries are two very similar concepts in Wasp. They even might appear to be doing the same thing, but Wasp treats them differently and using each construct carries certain semantic implications.
> 
> Here's a rundown of all the differences between Queries and Actions:
> 
> 1. Actions can (and most often should) modify the server's state, while Queries are only allowed to read it. Wasp relies on you following this convention when performing cache invalidations, so it's important to stick to it.
> 2. Actions don't need to be reactive so you can call them directly. Still, Wasp does provide a [`useAction` React hook](#the-useaction-hook) for adding extra behavior to the Action (e.g., optimistic updates).
> 3. `action` declarations in Wasp are mostly identical to `query` declarations. The only difference is in the declaration's name.

Reviewer: ```suggestion
Actions and Queries are two closely related concepts in Wasp. They might seem to perform similar tasks, but Wasp treats them differently, and each concept represents a different thing.

Here are the key differences between Queries and Actions:

1. Actions can (and often should) modify the server's state, while Queries are only permitted to read it. Wasp relies on you adhering to this convention when performing cache invalidations, so it's crucial to follow it.
2. Actions don't need to be reactive, so you can call them directly. However, Wasp does provide a [`useAction` React hook](#the-useaction-hook) for adding extra behavior to the Action (like optimistic updates).
3. `action` declarations in Wasp are mostly identical to `query` declarations. The only difference lies in the declaration's name.
```
